### PR TITLE
OFN DFC API documenation in OpenAPI format generated with Rswag

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

Related to:

- #9170

While I'm developing the DFC API, people ask me what it can and can't do. So it's time for some documentation to be open and save time in the end. I haven't actually seen any DFC app being documented in this standardised way. The next best thing is the [DFC documentation](https://datafoodconsortium.gitbook.io/dfc-standard-documentation/appendixes/practical-examples/version-1.8) which has only read examples, no update or create. And I don't know of any documentation which show the HTTP interaction with the API.

I noticed some shortcomings of the current API, which is still in development, but didn't fix them in this PR. For example:

- The SuppliedProducts#create action responds without content. It should respond with the created object, I guess. Otherwise you can't know its id.

But at least this setup of Rswag documents the *actual* behaviour instead of the expected behaviour. Rswag is usually supplied with JSON-Schema to specify what the API looks like and it can test the application against that. But I omitted that here because the DFC Connector is already enforcing that we use the right format and I didn't want to repeat the DFC ontology in a different format. Maybe we'll find a way to generate it automatically which could be helpful for the documentation.

Screen Shot 2023-06-28 at 14 38 55

#### What should we test?

- Visit `/api-docs` and have a browse.

#### Release notes

Changelog Category: Technical changes

The title of the pull request will be included in the release notes.

#### Dependencies

The documentation assumes that authentication via API token is possible already:

* https://github.com/openfoodfoundation/openfoodnetwork/pull/11101

#### Documentation updates

This is the documentation update. We could mention it somewhere but I don't know yet where people will be looking for it.

### CODE REVIEW

File: engines/dfc_provider/spec/requests/supplied_products_spec.rb [hunk 1]
```diff
@@ -1,91 +1,112 @@
...
+    let(:enterprise_id) { enterprise.id }
+
+    post "Create SuppliedProduct" do
+      consumes "application/json"
+      produces "application/json"
 
-    it "creates a variant" do
-      request_body = DfcLoader.connector.export(supplied_product)
+      # This parameter is required but I want to write a spec which doesn't
+      # supply it. I couldn't do it with rswag when requiring it.
```
Comment (dacook) [inline engines/dfc_provider/spec/requests/supplied_products_spec.rb:31]: Interesting, I suppose this feature needs to be added to rswag one day..

Or ideally, I wonder if  rswag could automatically add a test for a `400` when `required:true`? Either way, you'd want to be able to override it.

File: engines/dfc_provider/spec/support/example_json.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+module ExampleJson
+  def self.read(name)
+    pathname = DfcProvider::Engine.root.join("spec/support/#{name}.json")
```
Comment (dacook) [inline engines/dfc_provider/spec/support/example_json.rb:5]: Am I right to say that example files are normally called "fixtures"? In the main app, we actually have two fixtures folders:
1. `spec/fixtures/files/`
2. `spec/support/fixtures/`

Hmm, but also "fixtures" seems to be a more specific concept in ActiveRecord, for storing sample data in YAML in the `test/fixtures`. 

🤷 I think at least some dedicated directory would be a good idea to keep `spec/support/` more organised. My preference would be `#`1, what do you think?

**Review by dacook — APPROVED** [review]
Yay, this is exciting!!

Great job, I'm very happy to see this implemented, and done well in the one PR. 

I think it might be worth documenting when/why to run the `script/rswag` to keep it up to date (for both the API and DFC).

Also I have a thought about organising files..

File: spec/requests/api/v1/customers_spec.rb [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 require "swagger_helper"
 
-describe "Customers", type: :request do
+describe "Customers", type: :request, swagger_doc: "v1/swagger.yaml" do
```
[the remark this answers is not in the document]
Comment (mkllnk) [reply spec/requests/api/v1/customers_spec.rb:5]: Splitting up the swagger helpers is a nice idea. That may actually allow us to regenerate the files independently of each other. At the moment you have to run it all together. Maybe in a future PR...

File: engines/dfc_provider/spec/support/example_json.rb (hunk 2 above)
Comment (mkllnk) [reply engines/dfc_provider/spec/support/example_json.rb:5]: I was wondering that, too. These files are not ActiveRecord "fixtures". `spec/fixtures/files` makes that distinction and I would be okay with that. Currently, we have only three or four files in `spec/support` of that engine though. It's very tidy. When it becomes more, we can move it. And then we need to only change the ExampleJson module.

Comment (rioug) [reply engines/dfc_provider/spec/support/example_json.rb:5]: My preference would be 1. `spec/fixtures/files/` as well

**Review by rioug — APPROVED** [review]
Looks good !

**Pull request merged**